CustomOp simple abstract implementation registration

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #100127
* #100114
* __->__ #99439
* #100203

This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

zou3519 committed:
Simple FakeTensor registrations for custom ops

This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- adds a FakeTensor registration API for torch.library (impl_fake)

The FakeTensor implementation provided by the user:
- accepts FakeTensors and returns FakeTensors
- If the signature of the operator is (*args, **kwargs), then the
signature of the FakeTensor implementation is (ctx, *args, **kwargs).
- The `ctx` arg is a new FakeTensorImplCtx object that stores helper
methods. So far, the helper methods are (1) constructing a new unbacked
symint and (2) constraining the range of the symint.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 commented on torch/library.py:
```diff
@@ -149,6 +197,64 @@ def wrap(f):
     return wrap
 
 
+def impl_fake(lib, op_name, *, _stacklevel=3):
```
API design help wanted please: I'm not sure what the torch.library API for registering a FakeTensor implementation should be. Right now, the API is analogous to the existing torch.library.Library.impl and torch.library.impl, APIs in that they accept a torch.Library object.

However, we don't actually *need* the torch.library object, because the FakeTensor implementation isn't tied to a dispatch key! An alternate API (inspired by register_meta) could be look like the following:
```py
def impl_fake(op: OpOverload, ...): ...

@impl_fake(torch.ops.custom.foo)
def foo_fake(ctx, x):
    ...
```

If we prefer something more like this I can change it. The other considerations are that we will end up adding an `impl_autograd`, and `impl_autograd` may need to be tied to a torch.Library object.

ezyang replied:
So remember we were talking about the mid level vs high level API?

IMO, it doesn't make any sense to have fake tensor registrations in the mid level API, *unless* you want to add a fake tensor dispatch key (which is plausible, but we'd have to work out the consequences.) Mid level API is "stuff on top of existing torch.Library", tehrefore fake tensor shouldn't work here.

zou3519 committed:
Update on "CustomOp simple fake implementation registration"


This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The fake implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

eellison commented on torch/_custom_op.py:
```diff
@@ -370,3 +447,61 @@ def validate_function_matches_schema(
+# Used to query the CustomOp associated with a specific C++ dispatcher operator.
+# An example usage is FakeTensor: FakeTensor checks if a specific operator
+# has an implementation registered via the CustomOp API.
+global_registry: typing.Dict["qualname", "weakref.ref[CustomOp]"] = {}
```
weakref.WeakValueDictionary?

zou3519 replied:
Didn't know that existed, looks like that is what I am looking for. Thanks!

zou3519 committed:
Update on "CustomOp simple fake implementation registration"


This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The fake implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple fake implementation registration"


This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The fake implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple fake implementation registration"


This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The fake implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple fake implementation registration"


This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The fake implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

ezyang commented:
I support pubbing FakeTensor. Similarly, proxy_tensor and symbolic_shapes should graduate from experimental.

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -83,9 +86,116 @@ def numpy_take_impl(x, ind, ind_inv, dim):
+def numpy_take_fake(x, ind, ind_inv, dim):
+    return torch.empty_like(x)
+
+@custom_op('(Tensor x) -> Tensor', ns='_torch_testing')
```
I'm getting some buyer's remorse from removing the deduplication; the main downside is I can no longer grep _torch_testing::numpy_nonzero to find the operator that defines this lol

zou3519 replied:
Well, we can bring the name back and just assert that the name mentioned in the schema matches the function name:

```
# with schema
@custom_op('_torch_testing::numpy_nonzero', '(Tensor x) -> Tensor')
def numpy_nonzero(x):
    ...

# without schema (inferring schema from type annotations)
@custom_op('_torch_testing::numpy_nonzero')
def numpy_nonzero(x: Tensor) -> Tensor:
    ...
```

zou3519 replied:
(will leave for the future)

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -83,9 +86,116 @@ def numpy_take_impl(x, ind, ind_inv, dim):
+
+    ctx = torch._custom_op.get_ctx()
+    i0 = ctx.new_data_dependent_symint()
+    ctx.constrain_range(i0, min=2)
```
IMO the API should do this automatically

zou3519 replied:
Sounds good, there will just be a warning in the docs that symints must have minimum 2 otherwise things will go wrong.

We still probably want to provide ctx.constrain_range, for those people who have more of an idea what their dynamic dimension produces (e.g. an upper bound for nonzero is the total number of elements in the Tensor)

zou3519 replied:
Hmm, I take that back - we only want to constrain the range on symints that get used as the sizes of a Tensor. A user might not take `ctx.new_data_dependent_symint()` and directly pass it as a size to a new Tensor (though this is what most people who will use the API will do).

As an example, someone might have an op `foo(x, y)` that returns a tensor of shape [x_nnz + y_nnz]. One way to write the fake impl is:
```py
def foo_fake(x, y):
    ctx = get_ctx()
    x_nnz = ctx.new_data_dependent_symint()
    y_nnz = ctx.new_data_dependent_symint()
    out_nnz = ctx.constrain_range(x_nnz + y_nnz, 2)
    return x.new_empty([out_nnz])
```
This doesn't really matter right now because we don't do any memoization, but if we were to support memoization, then the user would actually have a reason to create two different symints (so they can be memoized separately)
```py
def foo_fake(x, y):
    ctx = get_ctx()
    x_nnz = ctx.new_data_dependent_symint(memoize_on=(foo_fake, x))
    y_nnz = ctx.new_data_dependent_symint(memoize_on=(foo_fake, y))
    out_nnz = ctx.constrain_range(x_nnz + y_nnz, 2)
    return x.new_empty([out_nnz])
```

zou3519 replied:
Resolved offline: we don't care enough, so going to constrain this to 2 by default.

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1247,6 +1248,16 @@ def dispatch(self, func, types, args=(), kwargs=None):
 
+        # Users can register FakeTensor rules for custom operators
+        # Call them if they exist.
+        if func.name() in torch._custom_op.global_registry:
```
Any reason this table isn't indexed on func itself?

zou3519 replied:
Lifetime issues, but I am very close to declaring bankruptcy on lifetime issues anyways.

The problem is that once we create the OpOverload, if the operator gets unregistered and re-registered, invoking the OpOverload leads to segfaults. So as written in this PR, we index the registry on qualname to avoid creating the OpOverload.

This only really matters for testing, but I might (in a future PR) just develop a mechanism for testing to clear custom OpOverloads because testing needs to materialize the OpOverloads anyways when we do things like run make_fx.

ezyang commented on torch/_custom_op.py:
```diff
@@ -370,3 +449,66 @@ def validate_function_matches_schema(
+        self._shape_env = _shape_env
+        self._op = _op
+
+    def new_data_dependent_symint(self) -> torch.SymInt:
```
docblocks for these?

ezyang commented on torch/_custom_op.py:
```diff
@@ -370,3 +449,66 @@ def validate_function_matches_schema(
+    return None
+
+
+global_ctx_getter: typing.Callable = get_none
```
Is there are particular reason this is a callable, as opposed to just holding the context itself?

ezyang commented on torch/_custom_op.py:
```diff
@@ -226,43 +232,116 @@ def inner(f):
+                        f"behavior. Otherwise, please remove the call to get_ctx() "
+                        f"in the implementation registered with impl_fake "
+                        f"at {new_location}"
+                    )
```
oh, I guess this is why

ezyang reviewed (approved):
okey dokey

zou3519 committed:
Update on "CustomOp simple fake implementation registration"


This PR:
- adds a FakeTensor registration API for CustomOp (CustomOp.impl_fake)
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The fake implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their FakeTensor
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.new_data_dependent_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -83,9 +86,114 @@ def numpy_take_impl(x, ind, ind_inv, dim):
+        raise RuntimeError("not supported")
+    return torch.tensor(res, device=x.device)
+
+@numpy_nonzero.impl_fake()
+def numpy_nonzero_fake(x):
+    ctx = torch._custom_op.get_ctx()
+    i0 = ctx.new_data_dependent_symint()
+    shape = [x.dim(), i0]
+    result = x.new_empty(shape, dtype=torch.long)
+    return result
```
Opening up the floor for some bikeshedding. Here's some questions to get started:

ctx.new_data_dependent_symint()
- ctx.new_symint()? ctx.new_size()? ctx.new_unbacked_symint()?

`numpy_nonzero.impl_fake`
- Should we call our "fake implementation" "abstract implementation" instead, and rename this to `impl_abstract`? "fake implementation" sounds a bit confusing to me because the implementation ends up being registered for both Meta tensors and FakeTensors, but another way to think about this is FakeTensor is a superset of Meta tensors.

ezyang replied:
Re (1), absent any pressure, I would line up the name with ShapeEnv, so `create_unbacked_symint`

Re (2), I actually kind of think we should lie and call it `impl_meta` anyway. I think my opinion is a bit different if we're going to support mixed-device custom ops, but if you're registering these implementations for Meta, they can't actually rely on a fake tensor mode being active (and thus always have to use `new_empty` and similar to make sure they are generated on the correct device.)

zou3519 replied:
> Re (1), absent any pressure, I would line up the name with ShapeEnv, so create_unbacked_symint

sgtm

> Re (2), I actually kind of think we should lie and call it impl_meta anyway. I think my opinion is a bit different if we're going to support mixed-device custom ops, but if you're registering these implementations for Meta, they can't actually rely on a fake tensor mode being active (and thus always have to use new_empty and similar to make sure they are generated on the correct device.)

I'm not sure we should lie, it would be bad if the user started creating new meta tensors inside the function (e.g. `torch.empty(3, 3, device=meta)`) because as it is currently set up: inside the FakeTensor rule, since the FakeTensorMode is active, we would interpret them as a FakeTensor wrapping a meta tensor

ezyang replied:
OK, well, then we shouldn't call it fake or meta lol.

zou3519 replied:
Why not fake? Because it's not always a FakeTensor rule? (e.g. when registered to meta?)

zou3519 replied:
> Re (2), I actually kind of think we should lie and call it impl_meta anyway. I think my opinion is a bit different if we're going to support mixed-device custom ops, but if you're registering these implementations for Meta, they can't actually rely on a fake tensor mode being active (and thus always have to use new_empty and similar to make sure they are generated on the correct device.)

So, one idea for this is that for the meta implementation, we can activate FakeTensor mode and send FakeTensors that wrap meta tensors into the implementation. The benefit of this is that we're no longer lying about this being a fake impl and that the user doesn't need to worry about a third thing (abstract impl vs meta impl vs fake impl).

zou3519 replied:
> So, one idea for this is that for the meta implementation, we can activate FakeTensor mode and send FakeTensors that wrap meta tensors into the implementation. The benefit of this is that we're no longer lying about this being a fake impl and that the user doesn't need to worry about a third thing (abstract impl vs meta impl vs fake impl).

This doesn't seem possible because the Meta dispatch key is after Python

zou3519 replied:
Updated to `impl_abstract`.

bdhirsh commented on torch/_custom_op.py:
```diff
@@ -370,3 +449,128 @@ def validate_function_matches_schema(
+
+            It is important that the ``min`` and ``max`` (if not None) values are set
+            correctly, otherwise, there will be undefined behavior under
+            torch.compile. The default value of ``min`` is 2.
```
Is it worth mentioning that the dfeault of 2 is due to 0/1 specialization? (is that right?)

zou3519 replied:
Yes. It's worth mentioning, I'll update it

bdhirsh commented on torch/_custom_op.py:
```diff
@@ -226,43 +232,116 @@ def inner(f):
+                with set_ctx_getter(error_on_ctx):
+                    return f(*args, **kwargs)
+
+            self._lib.impl(self._opname, f_with_ctx, "Meta")
```
question to cement my understanding: Is it the case that this registration to the Meta key will basically never be called when using torch.compile, since we're basically guaranteeing that every custom op will have both a Meta impl and a FakeTensor impl, and FakeTensorMode always gives priority to the FakeTensor impl?

I guess the only case where we'd use this meta impl is if there's user code directly using meta tensors.

bdhirsh commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -83,9 +86,114 @@ def numpy_take_impl(x, ind, ind_inv, dim):
+    assert scores.shape == (N,)
+
+    ctx = torch._custom_op.get_ctx()
+    i0 = ctx.new_data_dependent_symint()
```
very cool :)

bdhirsh replied:
I think you mentioned it before, but will it be possible to use this machinery to write FakeTensor rules for existing C++ custom ops?

bdhirsh commented on test/test_fake_tensor.py:
```diff
@@ -631,6 +634,22 @@ def contains_type(type: torch._C.Type, maybe_contained_type: torch._C.Type):
+                op(*args, **kwargs)
+            # This is allowed from a FakeTensor implementation. Dynamo knows
+            # how to graph break on them.
+            except DynamicOutputShapeException:
```
Is the point of these tests just to confirm that every custom op in the db either doesn't raise an exception, or raises a DynamicOutputShapeException? Is it worth tagging the customOpInfo's that will, and asserting that only the tagged ones will? (Not necessary if you don't think it's worth implementing).

zou3519 replied:
> Is the point of these tests just to confirm that every custom op in the db either doesn't raise an exception, or raises a DynamicOutputShapeException?

Yes. In addition to those, the CrossRef also checks that sizes/strides/devices are correct between the FakeTensor rule and actually executing the operators (for the ones that do not raise DynamicOutputShapeException)

> Is it worth tagging the customOpInfo's that will, and asserting that only the tagged ones will?

This is a good point. I'll add a simple tagging mechanism (i.e. a list somewhere) to do this.

eellison commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -83,9 +86,114 @@ def numpy_take_impl(x, ind, ind_inv, dim):
+    while order.size > 0:
+        i = order[0]
+        keep.append(i)
+        xx1 = np.maximum(x1[i], x1[order[1:]])
+        yy1 = np.maximum(y1[i], y1[order[1:]])
+        xx2 = np.minimum(x2[i], x2[order[1:]])
+        yy2 = np.minimum(y2[i], y2[order[1:]])
+
```
How does this work? We're running this on meta tensors here, so I would think this would throw

zou3519 committed:
Update on "CustomOp simple abstract implementation registration"


This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple abstract implementation registration"


This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple abstract implementation registration"


This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple abstract implementation registration"


This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple abstract implementation registration"


This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]

zou3519 committed:
Update on "CustomOp simple abstract implementation registration"


This PR:
- adds an abstract registration API for CustomOp (CustomOp.impl_abstract)
that is used for both FakeTensor and meta tensors
- deletes CustomOp.impl_meta

The user story behind this API is that it is the one-stop shop for
registering implementations for data-less Tensors, i.e. FakeTensor and
Meta tensor.

The abstract implementation provided by the user:
- gets registered as the FakeTensor implementation AND the meta formula
- can be written like a regular meta formula. If the user decides that
they need something more special (i.e. data-dependent output shape),
then they are able to query a current context object (FakeTensorImplCtx)
that has methods to construct new unbacked symints.

Caveats:
- we really need to make FakeTensor/FakeTensorMode public. Otherwise,
there isn't a way for the user to interactively test that their abstract
implementation is correct without running through large pieces of the
PT2 stack (make_fx or torch.compile).
- We do not memoize the symints produced by
ctx.create_unbacked_symint(). It is possible to do this in the
future, but it is difficult to do soundly and I am not convinced of
the utility outside of the nonzero() usecase mentioned in #95399

Public API:
- More docs will come when we actually expose this API to users by
putting it in a public namespace, unless you folks want it now.
- The APIs mentioned in `__all__` are the ones that are intended to be
public.

Test Plan:
- Updated existing custom_op_db operators
- Added new numpy_nonzero and numpy_nms operations that test operations
that have data-dependendent output shape.

[ghstack-poisoned]
